feat: Add local filesystem cloud storage adapter

realmeylisdev opened the pull request:
Closes #1817

## Summary

- Adds `serverpod_cloud_storage_local` package for storing files directly on the server filesystem
- No external cloud provider dependencies required (S3, GCP, etc.)
- Ideal for development, testing, and self-hosted deployments

## Features

- Store and retrieve files on local filesystem
- Public and private storage configurations
- Path sanitization to prevent directory traversal attacks
- Streaming support for memory-efficient large file handling
- Automatic expiration cleanup scheduler
- 51 unit tests covering core functionality

## Limitations

- Direct file uploads (`createDirectFileUploadDescription`) return `null`
- Use `storeFile()` to upload files through endpoints instead

## Usage

```dart
pod.addCloudStorage(
  LocalCloudStorage(
    serverpod: pod,
    storageId: 'public',
    storagePath: '/var/serverpod/uploads/public',
    public: true,
  ),
);
```

## Test plan

- [x] Unit tests pass (`dart test`)
- [ ] Manual testing with Serverpod project
- [ ] Review path sanitization for edge cases

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->

## Summary by CodeRabbit

## Release Notes

* **New Features**
  * Added a local filesystem cloud storage adapter for Serverpod enabling file storage and retrieval
  * Streaming support for efficient handling of large files
  * Automatic expiration-based cleanup scheduler for managing file lifecycles
  * Public and private storage configurations with customizable hosting options and path security safeguards

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

realmeylisdev committed:
feat: Add local filesystem cloud storage adapter

Adds a new `serverpod_cloud_storage_local` package that allows storing
files directly on the server's local filesystem instead of using cloud
storage providers like Amazon S3 or Google Cloud Storage.

Features:
- Store files on local filesystem with configurable storage path
- Support for public and private storage configurations
- Public URL generation for serving files through Serverpod's existing
  cloud storage endpoint
- Path sanitization to prevent directory traversal attacks
- File verification status tracking via metadata files

Limitations:
- Direct file uploads (createDirectFileUploadDescription) are not
  supported as they require database access for tracking upload entries.
  Users should use storeFile() through their endpoints instead.

This is useful for:
- Development and testing environments
- Self-hosted deployments
- Scenarios where external cloud dependencies are not desired

Closes #1817

realmeylisdev committed:
feat: Add streaming support, expiration cleanup, and tests for local storage

- Add storeFileStream/retrieveFileStream for memory-efficient large file handling
- Add expiration cleanup scheduler with startCleanupScheduler/stopCleanupScheduler
- Add getFileSize method for file size queries without loading content
- Improve path sanitization to prevent directory traversal attacks
- Add comprehensive test suite (51 tests)
- Update README with new features documentation

realmeylisdev committed:
chore: Fix formatting and pubspec dependency ordering

realmeylisdev committed:
fix: Update test dependency version to match other packages

realmeylisdev committed:
fix: Address CodeRabbit review feedback

- Relax serverpod version constraint to ^3.1.1 for semantic versioning
- Add path sanitization edge case tests for security coverage:
  - Multiple consecutive parent directory segments
  - Mixed separator traversal attempts
  - URL encoded traversal attempts
  - Special characters handling
  - Empty path default filename fallback

meylis1998 commented on integrations/serverpod_cloud_storage_local/lib/src/cloud_storage/local_cloud_storage.dart:
```diff
@@ -0,0 +1,564 @@
+    }
+
+    return p.join(storagePath, sanitized);
+  }
```
Addressed in commit 741a43d63. Added 6 new edge case tests covering:
- Multiple consecutive parent directory segments (`foo/../../../../../../etc/passwd`)
- Mixed separator traversal attempts (`foo\..\..\bar`)
- URL encoded traversal attempts (`foo%2F..%2Fbar`)
- Special characters handling
- Empty path default filename fallback

All 57 tests pass.

realmeylisdev committed:
fix: Use exact serverpod version for CI consistency

Change serverpod dependency from ^3.1.1 to 3.1.1 to match
all other packages in the monorepo.

nielsenko commented on integrations/serverpod_cloud_storage_local/LICENSE:
```diff
@@ -0,0 +1,21 @@
+Copyright 2020 The Serverpod authors
```
Should the year be updated?

nielsenko commented on integrations/serverpod_cloud_storage_local/README.md:
```diff
@@ -0,0 +1,158 @@
+A local filesystem storage adapter for Serverpod that allows you to store files directly on your server without requiring external cloud storage services like Amazon S3 or Google Cloud Storage.
+
+## Features
+
```
This entire file should be copied from README_subpackage.md

Make sure to update util/update_pubspec script to do so.

The extra information should be added to serverpod_docs repo. Maybe mention next to s3 and gcp storage in main repo (just a very short entry).

nielsenko commented on integrations/serverpod_cloud_storage_local/lib/src/cloud_storage/local_cloud_storage.dart:
```diff
@@ -0,0 +1,564 @@
+    // If empty after cleaning, use a default
+    if (sanitized.isEmpty) {
+      sanitized = 'file';
+    }
```
Huh? Why default to 'file'?

nielsenko commented on integrations/serverpod_cloud_storage_local/lib/src/cloud_storage/local_cloud_storage.dart:
```diff
@@ -0,0 +1,564 @@
+      final fullPath = _getFullPath(path);
+      final file = File(fullPath);
+
+      if (!file.existsSync()) {
```
use async version?
```suggestion
      if (!await file.exists()) {
```

nielsenko commented on integrations/serverpod_cloud_storage_local/lib/src/cloud_storage/local_cloud_storage.dart:
```diff
@@ -0,0 +1,564 @@
+    final fullPath = _getFullPath(path);
+    final file = File(fullPath);
+
+    if (!file.existsSync()) {
```
Why not await?

nielsenko reviewed (commented):
Could you elaborate a bit on flow regarding the `.meta` file?

realmeylisdev committed:
fix: Address PR review feedback for local cloud storage

- Move CHANGELOG entries to root, update util/update_pubspecs
- Use README_subpackage.md template
- Refactor path sanitization using path package utilities
- Use Uint8List.sublistView for ByteData conversion
- Replace sync existsSync() with async exists()

realmeylisdev commented:
Closing this in favor of #5305, which rebuilds the package from scratch on current `main` — this branch had conflicts with the new integrations structure (`s3_compat`/`r2`, generated pubspecs), and a fresh start made it possible to address all of the review feedback cleanly, including an explanation of the `.meta` file flow (@nielsenko's question) in the new PR description.

@Hanibachi Thanks for the suggestion on alternate data streams / xattrs! I stayed with portable sidecar files, since ADS/xattr support varies by filesystem and doesn't reliably survive copies, backups, and container volume mounts — but I've noted it in case the maintainers prefer that direction.
